sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRule.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2024-2024 Sermant Authors. All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package io.sermant.discovery.service.lb.rule;
+
+import com.alibaba.nacos.shaded.com.google.common.collect.Lists;
+
+import io.sermant.discovery.entity.ServiceInstance;
+
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+
+/**
+ * nacos Weight
+ *
+ * @author xiaozhao
+ * @since 2024-11-16
+ */
+public class NacosWeightRandomRule extends AbstractLoadbalancer {
+    @Override
+    protected ServiceInstance doChoose(String serviceName, List<ServiceInstance> instances) {
+        List<InstanceWithWeight> withWeights = instances.stream()
+                .map(instance -> {
+                    int weight = Integer.parseInt(instance.getMetadata().get("nacos.weight"));
+                    return new InstanceWithWeight(instance, weight);
+                }).collect(Collectors.toList());
+        return this.weightRandom(withWeights);
+    }
+
+    @Override
+    public String lbType() {
+        return "NacosWeight";
+    }
+
+    /**
+     * nacos Weight
+     *
+     * @author xiaozhao
+     * @since 2024-11-16
+     */
+    public static class InstanceWithWeight {
+        private ServiceInstance server;
+        private Integer weight;
+
+        /**
+         * constructor
+         *
+         * @param instance instance
+         * @param weight weight
+         */
+        public InstanceWithWeight(ServiceInstance instance, int weight) {
+            this.server = instance;
+            this.weight = weight;
+        }
+
+        public ServiceInstance getServer() {
+            return server;
+        }
+
+        public Integer getWeight() {
+            return weight;
+        }
+    }
+
+    /**
+     * 根据权重随机
+     * 算法参考 https://blog.csdn.net/u011627980/article/details/79401026
+     *
+     * @param list 实例列表
+     * @return 随机出来的结果
+     */
+    public ServiceInstance weightRandom(List<InstanceWithWeight> list) {
+        List<ServiceInstance> instances = Lists.newArrayList();
```

Comment:
please optimize this algorithm ，refer：
    int totalWeight = 0;
    for (InstanceWithWeight instanceWithWeight : list) {
        totalWeight += instanceWithWeight.getWeight();
    }
    int randomWeight = new Random().nextInt(totalWeight);
    int currentWeight = 0;
    for (InstanceWithWeight instanceWithWeight : list) {
        currentWeight += instanceWithWeight.getWeight();
        if (randomWeight < currentWeight) {
            return instanceWithWeight.getServer();
        }
    }